exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/efgs/OperationDao.java

```diff
@@ -1,161 +1,127 @@
+                "id", operation.operationId,
+                "batch_tag", operation.batchTag,
                 "keys_count_total", keysCountTotal,
                 "keys_count_201", keysCount201,
                 "keys_count_409", keysCount409,
-                "keys_count_500", keysCount500
+                "keys_count_500", keysCount500,
+                "updated_at", new Timestamp(Instant.now().toEpochMilli())
         )) == 1;
     }
 
-    public void markErrorOperation(long operationId) {
-        String sql = "update en.efgs_operation set state = cast(:error_state as en.state_t), updated_at = :updated_at, " +
-                "run_count = run_count + 1 " +
-                "where id = :id";
-        jdbcTemplate.update(sql, Map.of(
-                "error_state", EfgsOperationState.ERROR.name(),
-                "updated_at", new Timestamp(Instant.now().toEpochMilli()),
-                "id", operationId
-        ));
+    public void markErrorOperation(long operationId, Optional<String> batchTag) {
+        markErrorOperation(operationId, batchTag, new Timestamp(Instant.now().toEpochMilli()));
     }
 
-    public void setStalledToError() {
-        String sql = "update en.efgs_operation set state = cast(:error_state as en.state_t), updated_at = :updated_at " +
-                "where state = cast(:started_state as en.state_t) and (now() > (updated_at + interval '10 minute')) ";
-        jdbcTemplate.update(sql, Map.of(
-                "error_state", EfgsOperationState.ERROR.name(),
-                "started_state", EfgsOperationState.STARTED.name(),
-                "updated_at", new Timestamp(Instant.now().toEpochMilli())
-        ));
+    public void markErrorOperation(long operationId, Optional<String> batchTag, Timestamp timestamp) {
+        String sql = "update en.efgs_operation set state = cast(:error_state as en.state_t), updated_at = :updated_at, " +
+                "batch_tag = :batch_tag " +
+                "where id = :id";
+        Map<String, Object> params = new HashMap<>();
+        params.put("error_state", EfgsOperationState.ERROR.name());
+        params.put("updated_at", timestamp);
+        params.put("batch_tag", batchTag.orElse(null));
+        params.put("id", operationId);
+        jdbcTemplate.update(sql, params);
     }
 
-    public List<Long> getOutboundOperationsInError() {
-        String sql = "select id from en.efgs_operation " +
-                "where state = cast(:state as en.state_t) " +
-                "and direction = cast(:direction as en.direction_t) " +
-                "and run_count < :run_count";
+    public List<Timestamp> getAndResolveCrashed(EfgsOperationDirection direction) {
+        String sql = "update en.efgs_operation set state = cast(:new_state as en.state_t) " +
+                "where state = cast(:current_state as en.state_t) and direction = cast(:direction as en.direction_t) " +
+                "and updated_at < :max_age returning updated_at";
```

Comment:
max_age doesn't communicate very clearly that it's a timestamp (not age). Maybe stalled_operation_limit or somesuch?

Reply:
yep, I changed that